standalone/pruntime/enclave-api/src/lib.rs

```diff
@@ -150,3 +156,312 @@ pub mod blocks {
+            Counters {
+                next_block_number: self.sync_state.block_number_next,
+                next_header_number: self.sync_state.header_number_next,
+                next_para_header_number: 0,
+            }
+        }
+
+        pub fn sync_header(
+            &mut self,
+            headers: Vec<HeaderToSync>,
+            authority_set_change: Option<AuthoritySetChange>,
+        ) -> Result<chain::BlockNumber> {
+            self.sync_state
+                .sync_header(headers, authority_set_change, &mut self.state_roots)
+        }
+
+        pub fn feed_block(
+            &mut self,
+            block: &BlockHeaderWithEvents,
+            storage: &mut Storage,
+        ) -> Result<()> {
+            self.sync_state
+                .feed_block(block, &mut self.state_roots, storage)
+        }
+    }
+
+    #[derive(Default, Debug)]
+    pub struct Counters {
+        pub next_header_number: chain::BlockNumber,
+        pub next_para_header_number: chain::BlockNumber,
+        pub next_block_number: chain::BlockNumber,
+    }
+
+    pub struct ParachainSynchronizer<Validator> {
+        sync_state: BlockSyncState<Validator>,
+        last_relaychain_state_root: Option<Hash>,
+        para_header_number_next: chain::BlockNumber,
+        para_state_roots: VecDeque<Hash>,
+    }
+
+    impl<Validator: BlockValidator> ParachainSynchronizer<Validator> {
+        pub fn new(validator: Validator, main_bridge: u64) -> Self {
+            Self {
+                sync_state: BlockSyncState::new(validator, main_bridge),
+                last_relaychain_state_root: None,
+                para_header_number_next: 1,
+                para_state_roots: Default::default(),
+            }
+        }
+
+        pub fn counters(&self) -> Counters {
+            Counters {
+                next_block_number: self.sync_state.block_number_next,
+                next_header_number: self.sync_state.header_number_next,
+                next_para_header_number: self.para_header_number_next,
+            }
+        }
+
+        /// Given the relaychain headers in sequence, validate it and cached the last state_root for parachain header validation
+        pub fn sync_header(
+            &mut self,
+            headers: Vec<HeaderToSync>,
+            authority_set_change: Option<AuthoritySetChange>,
+        ) -> Result<chain::BlockNumber> {
+            let mut state_roots = Default::default();
+            let last_header =
+                self.sync_state
+                    .sync_header(headers, authority_set_change, &mut state_roots)?;
+            self.last_relaychain_state_root = state_roots.pop_back();
+            Ok(last_header)
+        }
+
+        /// Given the parachain headers in sequence, validate it and cached the state_roots for block validation
+        pub fn sync_parachain_header(
+            &mut self,
+            headers: Vec<chain::Header>,
+            proof: &[u8],
+            storage_key: &[u8],
```

Comment:
`storage_key` is always the key of `Paras.head(para_id)`. It can be passed from the outside, but should be created in the pRuntime, not from the relayer.

Reply:
Yes, it's created in pRuntime in the rpc handler.